pkg/ovsdb/monitors.go

```diff
@@ -5,198 +5,262 @@ import (
+				// remove itself
+				m.db.RemoveMonitor(m.dataBaseName)
+				return
+			}
+			result, _ := m.prepareTableUpdate(wresp.Events)
+			for hd, tu := range result {
+				go hd.handler.notify(hd.jsonValue, tu)
 			}
 		}
 	}()
-	return monitor
 }
 
-func (m *monitor) propagateEvent(event *clientv3.Event, destinations map[*updater][]*handlerKey) {
-	// for each updater entry create a return data and call each handler
+func (m *monitor) hasHandlers() bool {
+	return len(m.handlers) > 0
 }
 
-func (u *updater) prepareTableUpdate(event *clientv3.Event) (table string, uuid string, rowUpdate *ovsjson.RowUpdate, err error) {
-	key := string(event.Kv.Key)
-	if table, uuid, err = keyToTableAndUUID(key); err != nil {
-		return
+func mcrToUpdater(mcr ovsjson.MonitorCondRequest, isV1 bool) *updater {
+	sort.Strings(mcr.Columns)
+	var key string
+	for _, c := range mcr.Columns {
+		key = key + c
+	}
+	// TODO handle "Where"
+	if mcr.Select == nil {
+		mcr.Select = &allTrueSelect
+	}
+	key = fmt.Sprintf("%s%v%v", key, isV1, *mcr.Select)
+	return &updater{Columns: common.StringArrayToMap(mcr.Columns), isV1: isV1, Select: *mcr.Select, key: key}
+}
+
+func (m *monitor) prepareTableUpdate(events []*clientv3.Event) (map[handlerKey]ovsjson.TableUpdates, error) {
+	// prepare results that will be sent to clients
+	// handlerKey -> tableName -> uuid -> rowUpdate
+	result := map[handlerKey]ovsjson.TableUpdates{}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, ev := range events {
+		key := string(ev.Kv.Key)
+		path, table, uuid, err := keyToTableAndUUID(key)
+		if err != nil {
+			klog.Errorf("Wrong event's key %s", key)
+			continue
+		}
+		updaters, ok := m.key2Updaters[path]
+		if !ok {
+			klog.Infof("There is no monitors for table path %s", path)
+			continue
+		}
+		for _, updater := range updaters {
+			rowUpdate, err := updater.prepareRowUpdate(ev)
+			if err != nil {
+				klog.Errorf("prepareTableUpdate returned error %s, updater %v", err, updater)
+				continue
+			}
+			if rowUpdate == nil {
+				// there is no updates
+				continue
+			}
+			hKeys, ok := m.upater2handlers[updater.key]
+			if !ok {
+				klog.Errorf("Cannot find handlers for the updater %#v", updater)
+			}
+			for _, hKey := range hKeys {
+				tableUpdates, ok := result[hKey]
+				if !ok {
+					tableUpdates = ovsjson.TableUpdates{} //map[string]map[string]ovsjson.RowUpdate{}
+					result[hKey] = tableUpdates
+				}
+				tableUpdate, ok := tableUpdates[table]
+				if !ok {
+					tableUpdate = ovsjson.TableUpdate{} // map[string]ovsjson.RowUpdate{}
+					tableUpdates[table] = tableUpdate
+				}
+				// check if there is a rowUpdate for the same uuid
+				_, ok = tableUpdate[uuid]
+				if ok {
+					// TODO fixme
+					klog.Warningf("Duplicate event for %s/%s/%s", m.dataBaseName, table, uuid)
+				}
+				tableUpdate[uuid] = *rowUpdate
+			}
+		}
 	}
-	data := map[string]interface{}{}
+	return result, nil
+}
+
+func (u *updater) prepareRowUpdate(event *clientv3.Event) (*ovsjson.RowUpdate, error) {
```

Comment:
it is very difficult to verify that you have covered all cases as I don't see a clear structure here, you could for example due something along these lines (this is psaudo-code) using utility function to eliminate duplication.

```
switch event {
case "initial":
    switch select {
    case "insert":
        switch selectVersion {
        case "v1":
            switch columnType {
            case integer:
            case real:
            ...
            }
        case "v2":
        default:
            return errors.New("malformed select version")
        }
    case "delete"
    default:
        return errors.New("maformed select type")
    }
case "modify":
case "insert":
case "delete":
default:
    return errors.New("malformed event type")
}
```

Reply:
we cannot switch event, it has 2 methods isModified() and isCreated(), based on that we can separated different events, The version is boolean, so we don't need the default use case (and the switch)